[15.0][FIX] web_m2x_options did not work with widgets

CRogos opened the pull request:
There was a refactoring of the odoo JS code from v13->v14 which was not taken into account into the migration of the web_m2x_options.

This causes that the widget option did not work properly when web_m2x_options is installed. The red part is triggered by this widget and is missing. 
<field name="task_id" options="{'no_create_edit': True}" widget="task_with_hours"/>
![image](https://user-images.githubusercontent.com/1799080/195817214-cf25ae19-282f-42a3-9b91-1edb901067ec.png)

I've updated the search function from the [odoo/odoo](https://github.com/odoo/odoo/blob/15.0/addons/web/static/src/legacy/js/fields/relational_fields.js#L565) and refactored the implementation.

This was also an issue in the v14 of web_m2x_options so a backport might be useful when this change is merged.

CRogos commented:
Picked fix of #2304 and added configuration to sync (Create/Create+Edit) buttons in Dialog with dropdown (Create/Create+Edit) values.

I can squash the changes when reviewed.
@theangryangel could you have a look if this solution is working for you?

CRogos commented:
@nguyenhk99 , @goliveirab I saw you recently active on this module. Could please you review this fix?

CRogos commented:
Happy new year @pedrobaeza , could you have a look on this bug fixes. We are using it now for a while and it is working fine.

pedrobaeza commented:
I can't help too much on the JS side. Maybe @Tardo now that he is active again can help.

njeudy reviewed (approved):
Great, I just try and review code ! thanks

CRogos commented:
@sonhd91 could you please review this fix before migrating to v16?

CRogos commented:
@sanchonuria, @etobella, @kluna1998 could you also have a look on this fix. Same issue on v15 as https://github.com/OCA/web/pull/2409

etobella commented:
Well, it is not the same issue. The one fixed on 14.0 was related to M2ODialog, you try to fix the view and search. Isn't it?

CRogos commented:
@etobella, you are right. nevertheless would you be so kind and review this change? The bugfix is in our production since 4 months now, but here it is still not merged in master. 

I would like to port this module to Odoo 16, but first everything should be fixed.

CRogos commented:
@pedrobaeza do you have any advice how this PR can be processed faster? 
This PR is now open for 5 months and you might imagine that this is very frustrating.

pedrobaeza commented:
Yeah, I know, but the flows are there for ensuring certain stability in the changes that get the main repository. You can review other PRs, and ask in exchange that they review yours to speed up this process. You can also become a maintainer of the module (with enough merits), to exceptionally auto-merge these patches, making sure that the change is legit.

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -96,272 +101,250 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
+        }),
+
+        _canCreate: function () {
+            const self = this;
```
🔥 use 'this' directly

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -96,272 +101,250 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
+        },
+
+        _canShowQuickCreate: function () {
+            const self = this;
```
🔥 use 'this' directly

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -96,272 +101,250 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
+        },
+
+        _canShowCreateAndEdit: function () {
+            const self = this;
```
🔥  use 'this' directly

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -96,272 +101,250 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
+                    ir_options["web_m2x_options.create_edit"]
+                );
+
+            if (!create_edit_undef) return create_edit;
```
Please, use brackets

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -96,272 +101,250 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
+
+            if (!create_edit_undef) return create_edit;
+
+            if (!m2x_create_edit_undef) return m2x_create_edit;
```
Please, use brackets

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -96,272 +101,250 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
+                m2o_dialog_remove_buttons = is_option_set(
+                    ir_options["web_m2x_options.m2o_dialog_remove_buttons"]
+                );
+            if (!m2o_dialog_remove_buttons_undef) return m2o_dialog_remove_buttons;
```
Please, use brackets

Tardo reviewed (changes requested):
Thanks for the fix! 
For my part, only very minor changes are necessary.

I can't test it due to that runboat is not available. :/

CRogos commented:
@Tardo thank you for your feedback. I've updated the code and runboat is now also available.

Tardo reviewed (changes requested):
With the new changes the create/edit options are never shown:
![noex](https://user-images.githubusercontent.com/731270/228258486-62976567-fbfa-4297-8a47-a104e7ca356f.png)

It should show:
![m2x_act](https://user-images.githubusercontent.com/731270/228258628-305a8ecd-bcb1-4fc6-98cf-5efc77460429.png)

CRogos commented:
> With the new changes the create/edit options are never shown: ![noex](https://user-images.githubusercontent.com/731270/228258486-62976567-fbfa-4297-8a47-a104e7ca356f.png)
> 
> It should show: ![m2x_act](https://user-images.githubusercontent.com/731270/228258628-305a8ecd-bcb1-4fc6-98cf-5efc77460429.png)

Did you test on Runboat? 
In the System Parameters the config was set to 1 instead of True/False. I assume someone configured it manually?

![image](https://user-images.githubusercontent.com/1799080/228451253-96796e01-afdd-44fe-a3b3-c5d484afb0f9.png)

Could you retest and maybe install it on a clean database?

Tardo commented:
@CRogos It was me who put 1 instead of 'True' :S. In the master branch this configuration is not needed.
Am I missing something?

CRogos commented:
> @CRogos It was me who put 1 instead of 'True' :S. In the master branch this configuration is not needed. Am I missing something?

No you can delete these keys. Than the behaviour should be the Odoo default. But putting invalid/unexpected values in the configuration might cause an unexpected behavior. ;)

Or am I wrong and 1 was an valid value before my change?

Tardo commented on web_m2x_options/models/ir_config_parameter.py:
```diff
@@ -12,6 +12,7 @@ def get_web_m2x_options(self):
             "web_m2x_options.limit",
             "web_m2x_options.search_more",
             "web_m2x_options.m2o_dialog",
+            "web_m2x_options.m2o_dialog_remove_buttons",
```
Please, add documentation for this new parameter.

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -482,7 +476,7 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
             var limit = ir_options["web_m2x_options.field_limit_entries"];
             if (!_.isUndefined(limit)) {
-                attrs.limit = parseInt(limit);
+                attrs.limit = parseInt(limit, 10);
```
I prefer the use of "unary plus" as it is more concise with the operation we want to do here.
```suggestion
                attrs.limit = +limit;
```

CRogos replied:
Unary plus is not possible, because it is automatically replaced by pre-commit with:
Number(limit)

Tardo commented on web_m2x_options/static/src/js/form.js:
```diff
@@ -96,272 +107,255 @@ odoo.define("web_m2x_options.web_m2x_options", function (require) {
 
-        _search: function (search_val) {
-            var self = this;
+        _search: async function (searchValue = "") {
```
From my tests, this kind of statement doesn't work very well with inheritance. I'm not sure if it's still happening. I would prefer a more classic use of promises (new Promise...).

Tardo reviewed (changes requested):
The tests I have done on runbot are correct :)

Commit:
[IMP] Separate dialog button configuration
[IMP] Configure search_create_popup
M2ODialog does not function correctly due to up stream removal of search_create_popup custom_event, M2ODialog offers both Create and Create and Edit, even if the options are diabled
[FIX] web_m2x_options did not work with widgets

CRogos commented:
@Tardo this was a bigger refactoring as expected. Please re-test everything carefully.

Here are the changes compared with the previous PR: 
https://github.com/OCA/web/compare/adb8aeb..ca83e4950de4d97e115792c3ada856bf112b15c8

CRogos commented:
Not sure what to do with this PR. The module has now be migrated to v16 without these changes. As I will also switch to v16 soon, I will not do any effort into this issue anymore. 
Very frustrating that the issue could not be solved/merged in 6 months!
